packages/utils/src/lib/diff.ts

```diff
@@ -0,0 +1,59 @@
+export type Diff<T> = {
+  before: T;
+  after: T;
+};
+
+export function matchArrayItemsByKey<T>({
+  before,
+  after,
+  key,
+}: Diff<T[]> & { key: keyof T | ((item: T) => unknown) }) {
```

Comment:
By assigning these items `T[]`, what happens if they are actually undefined (groups, categories)? The `find` and other functions act as if they are always defined, or am I missing something?

Reply:
The function doesn't allow passing in an non-array value. The caller will be forced by TypeScript to do something like `?? []` when they pass in the argument.